Allow theming directory prompt completions

NikitaRevenco opened the pull request:
Currently, directories and files both use the same color. This makes it hard to distinguish between them. This PR adds the ability to pass custom styles to prompt items.

## Before

![image](https://github.com/user-attachments/assets/51aae193-acd1-4b83-a1eb-cf63ba05e86e)


## After
![image](https://github.com/user-attachments/assets/a0ce839f-32b1-4ee4-babc-3a378402b36c)

---

This PR adds `ui.text.directory` theme key for many themes. Screenshots of themes' base colors

<details>

<summary>
Screenshots of themes which are inherited from
</summary>

All themes in bullet points will show new colors for directories. 2nd level bullet point themes inherit from 1st level, so I only screenshot 1st level bullet point themes

- `catppuccin_mocha`
  - `catppuccin_latte`
  - `catppuccin_frappe`
  - `catppuccin_macchiato`

![image](https://github.com/user-attachments/assets/5e266ec7-f430-4af2-b9f9-14614b036833)

- `dark_plus`

![image](https://github.com/user-attachments/assets/8ab2f540-256f-4f0f-b98f-0e10cc38dfda)

- `dracula`

![image](https://github.com/user-attachments/assets/e36fbc1c-7b7c-4e90-89f5-470e5a1e7917)

- `dracula_at_night`

![image](https://github.com/user-attachments/assets/7e655410-0143-4c63-a6f7-95b1486c248b)

- `github_light`
  - `github_light_dimmed`
  - `github_light_colorblind`
  - `github_light_tritanopia`
  - `github_light_high_constrast`

![image](https://github.com/user-attachments/assets/93993102-5f51-483c-a5d1-56ff04e1ac0e)


- `github_dark`
  - `github_dark_dimmed`
  - `github_dark_colorblind`
  - `github_dark_tritanopia`
  - `github_dark_high_constrast`

![image](https://github.com/user-attachments/assets/b38a0db2-a3bd-4ead-8bfd-34f2fde149af)


- `gruvbox`
  - `gruvbox_light`
  - `gruvbox_light_hard`
  - `gruvbox_light_soft`
  - `gruvbox_dark_hard`
  - `gruvbox_dark_soft`

![image](https://github.com/user-attachments/assets/12f8a863-5985-49b7-98f5-4287ecb3a51c)


- `material_deep_ocean`
  - `material_darker`
  - `material_oceanic`
  - `material_palenight`

![image](https://github.com/user-attachments/assets/3e41aa83-15ab-4fa5-9326-f5263a0cf125)


- `rose_pine`
  - `rose_pine_dawn`
  - `rose_pine_moon`

![image](https://github.com/user-attachments/assets/64d63c33-a48c-4850-aacb-b69ebd1275e7)


- `tokyonight`
  - `tokyonight_day`
  - `tokyonight_moon`
  - `tokyonight_storm`

![image](https://github.com/user-attachments/assets/9f581bab-1f94-44e0-bbe1-d372a0b825f9)

- `default`

![image](https://github.com/user-attachments/assets/e599fb91-b3fd-4765-bb1d-c955a662d2cb)


</details>

archseer commented on helix-term/src/ui/mod.rs:
```diff
@@ -506,18 +508,28 @@ pub mod completers {
+        let style_from_file = |file: Cow<'_, str>| {
+            PathBuf::from_str(file.as_ref())
+                .ok()
+                .filter(|path| path.is_dir())
```
From what I understand this will trigger a syscall per entry. Directories already end with a slash/directory separator so you can just check the raw string for it

NikitaRevenco replied:
Thank you, I didn't know that! I've updated so it checks if the string ends with a slash instead

the-mikedavis commented on helix-term/src/ui/prompt.rs:
```diff
@@ -19,7 +20,8 @@ use helix_view::{
 type PromptCharHandler = Box<dyn Fn(&mut Prompt, char, &Context)>;
-pub type Completion = (RangeFrom<usize>, Cow<'static, str>);
+
+pub type Completion = (RangeFrom<usize>, Cow<'static, str>, Option<Style>);
```
Rather than passing a style as a separate enum member, this type should replace the `Cow<'static, str>` with a `Span<'static>` from the tui crate. It's effectively the same thing:

```rust
pub struct Span<'a> {
    pub content: Cow<'a, str>,
    pub style: Style,
}
```

NikitaRevenco replied:
Ok, I'm using `Span` instead of adding a 3rd tuple member now

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -506,18 +506,31 @@ pub mod completers {
+        let directory_color = editor.theme.get("function");
+
+        let style_from_file = |file: Cow<'_, str>| {
+            if file.ends_with('/') {
```
This should check for `std::path::MAIN_SEPARATOR` so it works on Windows as well

NikitaRevenco replied:
Ok, I did it

NikitaRevenco commented:
@the-mikedavis what color key should be used for the directories? Should I add `ui.text.directory`, for example or re-use something existing. Do I need to update each theme to add `ui.text.directory` in there?

NikitaRevenco commented:
Ok @the-mikedavis I've added `ui.text.directory` and added the colors for many themes, with screenshots for the themes which are *inherited from*:

- `catppuccin_mocha`
  - `catppuccin_latte`
  - `catppuccin_frappe`
  - `catppuccin_macchiato`

![image](https://github.com/user-attachments/assets/5e266ec7-f430-4af2-b9f9-14614b036833)

- `dark_plus`

![image](https://github.com/user-attachments/assets/8ab2f540-256f-4f0f-b98f-0e10cc38dfda)

- `dracula`

![image](https://github.com/user-attachments/assets/e36fbc1c-7b7c-4e90-89f5-470e5a1e7917)

- `dracula_at_night`

![image](https://github.com/user-attachments/assets/7e655410-0143-4c63-a6f7-95b1486c248b)

- `github_light`
  - `github_light_dimmed`
  - `github_light_colorblind`
  - `github_light_tritanopia`
  - `github_light_high_constrast`

![image](https://github.com/user-attachments/assets/93993102-5f51-483c-a5d1-56ff04e1ac0e)


- `github_dark`
  - `github_dark_dimmed`
  - `github_dark_colorblind`
  - `github_dark_tritanopia`
  - `github_dark_high_constrast`

![image](https://github.com/user-attachments/assets/b38a0db2-a3bd-4ead-8bfd-34f2fde149af)


- `gruvbox`
  - `gruvbox_light`
  - `gruvbox_light_hard`
  - `gruvbox_light_soft`
  - `gruvbox_dark_hard`
  - `gruvbox_dark_soft`

![image](https://github.com/user-attachments/assets/12f8a863-5985-49b7-98f5-4287ecb3a51c)


- `material_deep_ocean`
  - `material_darker`
  - `material_oceanic`
  - `material_palenight`

![image](https://github.com/user-attachments/assets/3e41aa83-15ab-4fa5-9326-f5263a0cf125)


- `rose_pine`
  - `rose_pine_dawn`
  - `rose_pine_moon`

![image](https://github.com/user-attachments/assets/64d63c33-a48c-4850-aacb-b69ebd1275e7)


- `tokyonight`
  - `tokyonight_day`
  - `tokyonight_moon`
  - `tokyonight_storm`

![image](https://github.com/user-attachments/assets/9f581bab-1f94-44e0-bbe1-d372a0b825f9)

the-mikedavis commented on helix-term/src/ui/mod.rs:
```diff
@@ -506,18 +507,30 @@ pub mod completers {
+        let directory_color = editor.theme.get("ui.text.directory");
+
+        let style_from_file = |file: Cow<'static, str>| {
+            if file.ends_with(std::path::MAIN_SEPARATOR) {
```
Instead of checking the end of the string we can actually determine if these files are directories or without the filesystem-access-per-file (mentioned earlier in review) by uncommenting from above:

```rust
let is_dir = entry.file_type().is_some_and(|entry| entry.is_dir());
```

(`DirEntry` already stores the file type and `file_type` just accesses that info, no extra fs access.)

This should fix the colors when using something like `:cd` that uses the directory completer.

There's a bit of overlap between this and #11285 so I would suggest adding a struct near the top of this file

```rust
struct Utf8PathBuf {
    path: String,
    is_dir: bool,
}
```

And the `filter_map` in `filename_impl` can return that type. Then to make it play well with `fuzzy_match` helper we can implement `AsRef<str>` for that struct:

```rust
impl AsRef<str> for Utf8PathBuf {
    fn as_ref(&str) -> &str {
        &self.path
    }
}
```

Then later #11285 can reuse this type instead of the `(PathBuf, bool)` it uses currently.

NikitaRevenco replied:
thanks, added in https://github.com/helix-editor/helix/pull/12205/commits/b45d3ebb05cb022529cb97a2d2c13defcf234348

the-mikedavis commented on helix-term/src/ui/prompt.rs:
```diff
@@ -446,18 +449,24 @@ impl Prompt {
+                    completion.style
                 } else {
-                    completion_color
+                    default_completion_color
```
Here you can use `completion_color.patch(completion.style)` for roughly the same effect

NikitaRevenco replied:
neat! Added in https://github.com/helix-editor/helix/pull/12205/commits/cefa10da700c8712d2f9d444c8eea89f32b688f4

NikitaRevenco commented on helix-term/src/ui/mod.rs:
```diff
@@ -306,7 +307,7 @@ pub mod completers {
         fuzzy_match(input, names, false)
             .into_iter()
-            .map(|(name, _)| ((0..), name.into()))
+            .map(|(name, _)| ((0..), Span::raw(name)))
```
reverted these in https://github.com/helix-editor/helix/pull/12205/commits/0f254ee5413e06fb8f4e016ef39db7dd43f055c1

David-Else commented:
I am the maintainer of `dark_plus` and I approve of your color choice, thanks!

the-mikedavis reviewed (approved):
Thanks!
